proof/refine/ARM/Ipc_R.thy

```diff
@@ -3156,6 +3156,461 @@ lemma maybeReturnSc_corres:
+         apply (clarsimp simp: valid_sched_context_def list_all_def obj_at_def)
+        apply clarsimp
+        apply (case_tac "sc_replies x"; simp)
+        apply (intro allI impI conjI)
+        apply (subgoal_tac "replyPrevs_of s a = replyPrev ko")
+         apply (erule rsubst2[where P="\<lambda>a b. heap_list a b c" for c])
+          apply (rule ext, simp)
+         apply (simp)
+        apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def projectKO_reply opt_map_def)
+       apply assumption
+      apply (clarsimp simp: obj_at_def  is_sc_obj)
+      apply (frule state_relation_sc_replies_relation)
+      apply (subgoal_tac "scReply sc = hd_opt (sc_replies sca)")
+       apply (intro conjI)
+          apply (clarsimp)
+         apply (clarsimp)
+         apply (erule (1) reply_at_cross[rotated])
+          apply (frule (1) valid_sched_context_objsI)
+          apply (clarsimp simp: valid_sched_context_def list_all_def obj_at_def)
+         apply fastforce
+        apply (drule_tac x=scptr in spec)
+        apply (clarsimp simp: sc_at_pred_n_def obj_at_def)
+       apply (erule (1) sc_replies_relation_prevs_list)
+       apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def projectKO_sc)
+      apply (erule (1) sc_replies_relation_scReply[symmetric])
+      apply (clarsimp simp: obj_at'_real_def  ko_wp_at'_def projectKO_sc)
+     apply wpsimp+
+  done
+
+lemma valid_tcb_state_BlockedOnReply[simp]:
+  "valid_tcb_state (Structures_A.thread_state.BlockedOnReply reply_ptr) = reply_at reply_ptr"
+  by (rule ext, simp add: valid_tcb_state_def)
+
+crunches setReplyTCB
+  for valid_tcbs'[wp]: valid_tcbs'
+
+(* this is a forwards-reasoning rule for cross_rel *)
+lemma cross_relF:
+  "(s, s') \<in> state_relation \<Longrightarrow> cross_rel A B \<Longrightarrow> A s \<Longrightarrow> B s'"
+  by (clarsimp simp: cross_rel_def)
+
+lemma reply_at'_cross_rel:
+  "cross_rel (pspace_aligned and pspace_distinct and reply_at t) (reply_at' t)"
+  unfolding cross_rel_def state_relation_def
+  apply clarsimp
+  by (erule (3) reply_at_cross)
+
+lemma valid_tcbs_valid_tcbE':
+  assumes "valid_tcbs' s"
+          "tcb_at' t s"
+          "\<And>tcb. ko_at' tcb t s \<Longrightarrow> valid_tcb' tcb s \<Longrightarrow> R s tcb"
+  shows "obj_at' (R s) t s"
+  using assms
+  apply (clarsimp simp add: projectKOs valid_tcbs'_def ran_def typ_at'_def
+                            ko_wp_at'_def valid_obj'_def valid_tcb'_def obj_at'_def)
+  done
+
+lemma replyPush_corres:
+  "can_donate = can_donate' \<Longrightarrow>
+   corres dc (invs and tcb_at caller and tcb_at callee and reply_at reply_ptr
+              and (\<lambda>s. \<forall>x. sc_replies_sc_at (\<lambda>a. reply_ptr \<notin> set a) x s))
+             (valid_tcbs' and valid_release_queue and valid_release_queue')
+   (reply_push caller callee reply_ptr can_donate)
+   (replyPush caller callee reply_ptr can_donate')"
+  unfolding reply_push_def replyPush_def
+  apply (clarsimp simp: )
+  apply add_sym_refs
```

Comment:
I don't remember the exact details, but it might also help to use `add_valid_replies` here.